### PR TITLE
[0011] add transaction filter rfc

### PR DESCRIPTION
**Pull request by quake**
- [x] Transaction with index schema definition
- [x] Add link to serialization RFC

### CODE REVIEW

File: rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+These peers do not attempt to fully verify the block chain, instead just checking that [block headers connect](../0004-ckb-block-sync/0004-ckb-block-sync.md#connecting-header) together correctly and trusting that the transactions in the block of high difficulty are in fact valid.
+
+Without this protocol, peers have to download the entire blocks and accept all broadcast transactions, then throw away majority of the transactions. This slows down the synchronization process, wastes users bandwidth and increases memory usage.
+
+## Messages
+
+### SetFilter
+
+Upon receiving a `SetFilter` message, the remote peer will immediately restrict the broadcast transactions it announces to transactions matching the filter, where the [matching algorithm](#filter-matching-algorithm) is specified below.
```
Comment (doitian) [inline rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md]: The sentence is a bit too complex.

```suggestion
Upon receiving a `SetFilter` message, the remote peer will immediately restrict the transactions that it broadcasts to the ones matching the filter, where the [matching algorithm](#filter-matching-algorithm) is specified below.
```

File: rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+`hashes`: Partial Merkle branch proof.
+
+## Filter matching algorithm
+
+The filter can be tested against all broadcast transactions, to determine if a transaction matches the filter, the following algorithm is used. Once a match is found the algorithm aborts.
+
+1. Test the hash of the transaction itself.
+2. For each CellInput, test the hash of `previous_output` and `type hash` of unlock script.
+3. For each CellOutput, test the `lock` and `type hash` of contract.
```
Comment (doitian) [inline rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md]: @xxuejie plz take a look at this

File: rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    transactions: [[Transaction, uint32]];
+    hashes: [H256];
+}
+```
+
+`header`: Standard block header struct.
+
+`transactions`: Standard transaction struct plus transaction index.
+
+`hashes`: Partial Merkle branch proof.
```
Comment (doitian) [inline rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md]: It's better to add a link to static Merkle tree RFC?

File: rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+## Abstract
+
+Transaction filter protocol allows peers to reduce the amount of transaction data they are sent. Peer that wants to retrieve transactions of interest, has the option of setting filters on each connection. A filter is defined as a [Bloom filter](http://en.wikipedia.org/wiki/Bloom_filter) on data derived from transactions.
+
+## Motivation
+
+The purpose of transaction filter protocol is to allow low-capacity peers (smartphones, browser extensions, embedded devices, etc) to maintain a high-security assurance about the up to date state of some particular transactions of the chain or verify the execution of transactions.
+
+These peers do not attempt to fully verify the block chain, instead just checking that [block headers connect](../0004-ckb-block-sync/0004-ckb-block-sync.md#connecting-header) together correctly and trusting that the transactions in the block of high difficulty are in fact valid.
```
Comment (xxuejie) [inline rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md]: `high difficulty` -> `highest difficulty`?

File: rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md (hunk 2 above)
Comment (xxuejie) [reply rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md]: Those seem a good list to me. One additional question is: is it true that we only implement a filter that matches a transaction as long as any of the fields here(such as `type hash`, `lock`, etc) matches the filter? In other words, will we implement some sort of filter that can just test for a single specified field, such as `lock`?

File: rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+### AddFilter
+
+Upon receiving a `AddFilter` message, the given bit data will be added to the exsiting filter via bitwise OR operator. A filter must have been previously provided using `SetFilter`. This messsage is useful if a new filter is added to a peer whilst it has connections to the network open, it avoids the need to re-calculate and send an entirely new filter to every peer.
+
+```
+table AddFilter {
+    filter: [uint8];
+}
+```
```
Comment (zhangsoledad) [inline rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md]: Shall we declare which serialize format will be used？

File: rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md (hunk 2 above)
Comment (quake) [reply rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md]: > Those seem a good list to me. One additional question is: is it true that we only implement a filter that matches a transaction as long as any of the fields here(such as `type hash`, `lock`, etc) matches the filter? In other words, will we implement some sort of filter that can just test for a single specified field, such as `lock`?

Bloom filter is a probabilistic data structure, peer needs to check or filter transaction again after receiving it, so let's keep protocol simple.

File: rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md (hunk 5 above)
Comment (quake) [reply rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md]: will link to https://github.com/nervosnetwork/rfcs/pull/47 when it's accepted.

Comment (doitian) [reply rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md]: Link to this https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0008-serialization/0008-serialization.md

File: rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+
+### FilteredBlock
+
+After a filter has been set, peers don't merely stop announcing non-matching transactions, they can also serve filtered blocks. This message is a replacement for `Block` message of sync protocol and `CompactBlock` message of relay protocol.
+
+```
+table FilteredBlock {
+    header: Header;
+    transactions: [[Transaction, uint32]];
```
Comment (doitian) [inline rfcs/0000-transaction-filter-protocol/0000-transaction-filter-protocol.md]: Not a valid FlatBuffers type.

**Pull request merged**